circleguard/visualizer.py

```diff
@@ -80,219 +129,242 @@ def next_frame_from_timer(self):
+            hit_t = current_hitobj.time.total_seconds() * 1000
+            if isinstance(current_hitobj, Slider) or isinstance(current_hitobj, Spinner):
+                hit_end = self.get_hit_endtime(current_hitobj) + (self.fade_in)
+            else:
+                hit_end = hit_t + self.hitwindow + (self.fade_in)
+            if hit_t - self.preempt < current_time < hit_end:
                 self.hitobjs.append(current_hitobj)
-            elif current_hitobj.time-preempt > time and not 2 & current_hitobj.type:
+            elif hit_t > current_time:
                 found_all = True
-            index += 1
-            if index == len(self.beatmap.hitobjects)-1:
+            if index == len(self.beatmap.hit_objects) - 1:
                 found_all = True
-
+            index += 1
 
     def paintEvent(self, event):
         """
         Called whenever self.update() is called. Draws all cursors and Hitobjects
         """
-        painter = QPainter(self)
-        painter.setRenderHint(QPainter.Antialiasing)
+        self.painter.begin(self)
+        self.painter.setRenderHint(QPainter.TextAntialiasing, True)
+        self.painter.setRenderHint(QPainter.Antialiasing, True)
+        self.painter.setRenderHint(QPainter.HighQualityAntialiasing, True)
         black_bg = get_setting("visualizer_bg")
-        painter.setPen(PEN_WHITE if (get_setting("dark_theme") or black_bg) else PEN_BLACK)
+        self.painter.setPen(PEN_WHITE if (get_setting("dark_theme") or black_bg) else PEN_BLACK)
+        _pen = self.painter.pen()
         if black_bg:
             pal = QPalette()
             pal.setColor(QPalette.Background, Qt.black)
             self.setAutoFillBackground(True)
             self.setPalette(pal)
+        # loading screen
+        if self.loading_flag:
+            if self.thread.is_alive():
+                self.draw_loading_screen()
+                self.painter.end()
+                return
+            else:
+                self.loading_flag = False
+                self.clock.reset()
+                self.painter.end()
+                return
+        # debug stuff
+        self.frame_times.insert(0, self.frame_time_clock.get_time() - self.last_frame)
+        self.frame_times = self.frame_times[:120]
+        self.last_frame = self.frame_time_clock.get_time()
+        # beatmap
+        if self.beatmap_flag:
+            self.paint_beatmap()
+        # cursors
+        for player in self.players:
+            self.paint_cursor(player)
+        # other info
+        self.painter.setPen(_pen)
         if get_setting("visualizer_info"):
-            self.paint_info(painter)
-        if self.beatmap_path != "":
-            self.paint_beatmap(painter)
-        for index in range(self.replay_amount):
-            self.paint_cursor(painter, index)
+            self.paint_info()
+        if get_setting("visualizer_frametime"):
+            self.paint_frametime_graph()
+        self.painter.end()
```

Comment:
is this necessary? docs don't suggest calling it: https://doc.qt.io/qt-5/qpainter.html#end

Reply:
Removing it just instantly crashes on the second frame. This is most likely necessary since we reuse the painter.